cloudinit/config/cc_wireguard.py

```diff
@@ -0,0 +1,291 @@
+    """Writing user-provided configuration into Wireguard
+    interface configuration file.
+
+    @param wg_int: Dict of configuration value under 'wg:interfaces'.
+
+    @raises: RuntimeError for issues writing of configuration file.
+    """
+    LOG.debug("Configuring Wireguard interface %s", {wg_int["name"]})
+    try:
+        LOG.debug(
+            "Writing wireguard config to file %s", {wg_int["config_path"]}
+        )
+        util.write_file(
+            wg_int["config_path"], wg_int["content"], mode=WG_CONFIG_FILE_MODE
+        )
+    except Exception as e:
+        raise RuntimeError(
+            "Failure writing Wireguard configuration file"
+            f' {wg_int["config_path"]}:{NL}{str(e)}'
+        ) from e
+
+
+def enable_wg(wg_int: dict, cloud: Cloud):
+    """Enable and start Wireguard interface
+
+    @param wg_int: Dict of configuration value under 'wg:interfaces'.
+
+    @raises: RuntimeError for issues enabling WG interface.
+    """
+    try:
+        LOG.debug("Enabling wg-quick@%s at boot", {wg_int["name"]})
+        cloud.distro.manage_service("enable", f'wg-quick@{wg_int["name"]}')
+        LOG.debug("Bringing up interface wg-quick@%s", {wg_int["name"]})
+        cloud.distro.manage_service("start", f'wg-quick@{wg_int["name"]}')
+    except subp.ProcessExecutionError as e:
+        raise RuntimeError(
+            f"Failed enabling/starting Wireguard interface(s):{NL}{str(e)}"
+        ) from e
+
+
+def readinessprobe_command_validation(wg_readinessprobes: list):
+    """Basic validation of user-provided probes
+
+    @param wg_readinessprobes: List of readinessprobe probe(s).
+
+    @raises: ValueError of wrong datatype provided for probes.
+    """
+    errors = []
+    pos = 0
+    for c in wg_readinessprobes:
+        if not isinstance(c, str):
+            errors.append(
+                f"Expected a string for readinessprobe at {pos}. Found {c}"
+            )
+            pos += 1
+
+    if errors:
+        raise ValueError(
+            f"Invalid readinessProbe commands:{NL}{NL.join(errors)}"
+        )
+
+
+def readinessprobe(wg_readinessprobes: list):
+    """Execute provided readiness probe(s)
+
+    @param wg_readinessprobes: List of readinessprobe probe(s).
+
+    @raises: ProcessExecutionError for issues during execution of probes.
+    """
+    errors = []
+    for c in wg_readinessprobes:
+        try:
+            LOG.debug("Running readinessprobe: '%s'", {str(c)})
+            subp.subp(c, capture=True, shell=True)
+        except subp.ProcessExecutionError as e:
+            errors.append(f"{c}: {e}")
+
+    if errors:
+        raise RuntimeError(
+            f"Failed running readinessprobe command:{NL}{NL.join(errors)}"
+        )
+
+
+def maybe_install_wireguard_packages(cloud: Cloud):
+    """Install wireguard packages and tools
+
+    @param cloud: Cloud object
+
+    @raises: Exception for issues during package
+    installation.
+    """
+
+    packages = ["wireguard-tools"]
```

Comment:
On Ubuntu I think we may need to install `resolvconf` for normal mode of operation on cloud-init, since `wg-quick` calls resolvconf, but it isn't guaranteed to be installed, since resolvconf isn't listed as a package dependency of wireguard (only suggested).

Reply:
I suppose someone might want to use wireguard without dns settings, therefore not needing resolvconf. Feel free to ignore this suggestion.

@chifac08 Do you think it is worth mentioning that resolvconf would be required for dns in the docs? Let me know your thoughts.

Reply:
@holmanb i think we should document it but not install `resolvconf` on default. As most modern Linux distributions use systemd for DNS - `resolvectl` the user should be responsible for installing `resolvconf` package or just symlink it to `resolvectl`. I am going to add this to documentation.

Reply:
@chifac08 Thanks, that makes sense to me!